File: native/bindings/endstone/map/map.cpp
#include "map.h"

#include "bindings/endstone/server.h"

#include <aegilex-runtime/src/cxx_host_ui.rs.h>

#include <endstone/level/dimension.h>
#include <endstone/level/level.h>
#include <endstone/map/map_view.h>
#include <endstone/server.h>

#include <optional>
#include <string>
#include <string_view>

namespace aegilex::native::ui {

class Map::impl {
  public:
    explicit impl(endstone::MapView *map, endstone::Server *server) noexcept : map(map), server(server)
    {
    }

    endstone::MapView *map;
    endstone::Server *server;
};

namespace {

[[nodiscard]] std::optional<endstone::MapView::Scale> to_endstone_scale(const std::uint8_t scale) noexcept
{
    switch (scale) {
    case 0:
        return endstone::MapView::Scale::Closest;
    case 1:
        return endstone::MapView::Scale::Close;
    case 2:
        return endstone::MapView::Scale::Normal;
    case 3:
        return endstone::MapView::Scale::Far;
    case 4:
        return endstone::MapView::Scale::Farthest;
    }
    return std::nullopt;
}

} // namespace

Map::Map(endstone::MapView *map, endstone::Server *server) noexcept
    : impl(std::make_shared<class Map::impl>(map, server))
{
}

std::int64_t Map::getId() const
{
    try {
        return impl->map->getId();
    }
    catch (...) {
        return 0;
    }
}

bool Map::isVirtual() const
{
    try {
        return impl->map->isVirtual();
    }
    catch (...) {
        return false;
    }
}

std::uint8_t Map::getScale() const
{
    try {
        return static_cast<std::uint8_t>(impl->map->getScale());
    }
    catch (...) {
        return 0;
    }
}

void Map::setScale(const std::uint8_t scale) const
{
    try {
        const auto map_scale = to_endstone_scale(scale);
        if (!map_scale) {
            return;
        }
        impl->map->setScale(*map_scale);
    }
    catch (...) {
    }
}

std::int32_t Map::getCenterX() const
{
    try {
        return impl->map->getCenterX();
    }
    catch (...) {
        return 0;
    }
}

void Map::setCenterX(const std::int32_t x) const
{
    try {
        impl->map->setCenterX(x);
    }
    catch (...) {
    }
}

std::int32_t Map::getCenterZ() const
{
    try {
        return impl->map->getCenterZ();
    }
    catch (...) {
        return 0;
    }
}

void Map::setCenterZ(const std::int32_t z) const
{
    try {
        impl->map->setCenterZ(z);
    }
    catch (...) {
    }
}

rust::String Map::getDimensionName() const
{
    try {
        const auto *dimension = impl->map->getDimension();
        return rust::String(dimension != nullptr ? dimension->getName() : "");
    }
    catch (...) {
        return rust::String();
    }
}

bool Map::isUnlimitedTracking() const
{
    try {
        return impl->map->isUnlimitedTracking();
    }
    catch (...) {
        return false;
    }
}

void Map::setUnlimitedTracking(const bool unlimited) const
{
    try {
        impl->map->setUnlimitedTracking(unlimited);
    }
    catch (...) {
    }
}

bool Map::isLocked() const
{
    try {
        return impl->map->isLocked();
    }
    catch (...) {
        return false;
    }
}

void Map::setLocked(const bool locked) const
{
    try {
        impl->map->setLocked(locked);
    }
    catch (...) {
    }
}

bool Map::setDimension(const rust::Str dimension) const
{
    try {
        const std::string_view dimension_value(dimension.data(), dimension.size());
        if (impl == nullptr || impl->map == nullptr || impl->server == nullptr) {
            return false;
        }
        auto *level = impl->server->getLevel();
        auto *target = level == nullptr ? nullptr : level->getDimension(std::string(dimension_value));
        if (target == nullptr) {
            return false;
        }
        impl->map->setDimension(*target);
        return true;
    }
    catch (...) {
        return false;
    }
}

endstone::MapView *Map::native() const noexcept
{
    return impl == nullptr ? nullptr : impl->map;
}

} // namespace aegilex::native::ui

File: rust/crates/aegilex-runtime/src/host/endstone/map/map.rs
//! Core ABI implementation for `native/bindings/endstone/map/map.h`.

use crate::host::endstone::support::*;

impl crate::core_host::imports::HostMapView for PluginStoreState {
    fn exists(&mut self, id: i64) -> Result<Result<bool, TypesHostError>, String> {
        Ok(Ok(resolve_map_native(self, id).is_ok()))
    }

    fn create(&mut self, dimension: String) -> Result<Result<i64, TypesHostError>, String> {
        Ok((|| {
            let map = self
                .host
                .server()
                .map_err(map_core_host_error)?
                .createMap(&dimension);
            (!map.is_null())
                .then(|| map.getId())
                .ok_or_else(|| map_core_host_error(not_found()))
        })())
    }

    fn get_scale(&mut self, id: i64) -> Result<Result<MapViewMapScale, TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .and_then(|map| match map.getScale() {
                0 => Ok(MapViewMapScale::Closest),
                1 => Ok(MapViewMapScale::Close),
                2 => Ok(MapViewMapScale::Normal),
                3 => Ok(MapViewMapScale::Far),
                4 => Ok(MapViewMapScale::Farthest),
                _ => Err(invalid_input()),
            })
            .map_err(map_core_host_error))
    }

    fn set_scale(
        &mut self,
        id: i64,
        scale: MapViewMapScale,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            let scale = match scale {
                MapViewMapScale::Closest => 0u8,
                MapViewMapScale::Close => 1u8,
                MapViewMapScale::Normal => 2u8,
                MapViewMapScale::Far => 3u8,
                MapViewMapScale::Farthest => 4u8,
            };
            resolve_map_native(self, id)
                .map(|map| map.setScale(scale))
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn get_center_x(&mut self, id: i64) -> Result<Result<i32, TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .map(|map| map.getCenterX())
            .map_err(map_core_host_error))
    }

    fn set_center_x(&mut self, id: i64, x: i32) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            resolve_map_native(self, id)
                .map(|map| map.setCenterX(x))
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn get_center_z(&mut self, id: i64) -> Result<Result<i32, TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .map(|map| map.getCenterZ())
            .map_err(map_core_host_error))
    }

    fn set_center_z(&mut self, id: i64, z: i32) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            resolve_map_native(self, id)
                .map(|map| map.setCenterZ(z))
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn get_dimension(&mut self, id: i64) -> Result<Result<String, TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .map(|map| map.getDimensionName())
            .map_err(map_core_host_error))
    }

    fn set_dimension(
        &mut self,
        id: i64,
        dimension: String,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .and_then(|map| {
                map.setDimension(&dimension)
                    .then_some(())
                    .ok_or_else(not_found)
            })
            .map_err(map_core_host_error))
    }

    fn is_virtual(&mut self, id: i64) -> Result<Result<bool, TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .map(|map| map.isVirtual())
            .map_err(map_core_host_error))
    }

    fn is_unlimited_tracking(&mut self, id: i64) -> Result<Result<bool, TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .map(|map| map.isUnlimitedTracking())
            .map_err(map_core_host_error))
    }

    fn set_unlimited_tracking(
        &mut self,
        id: i64,
        unlimited: bool,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            resolve_map_native(self, id)
                .map(|map| map.setUnlimitedTracking(unlimited))
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn is_locked(&mut self, id: i64) -> Result<Result<bool, TypesHostError>, String> {
        Ok(resolve_map_native(self, id)
            .map(|map| map.isLocked())
            .map_err(map_core_host_error))
    }

    fn set_locked(&mut self, id: i64, locked: bool) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            resolve_map_native(self, id)
                .map(|map| map.setLocked(locked))
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }
}
